### ISSUE TITLE
Spring-boot 1.3.0.* - Autowired configuration not loaded before dependent configuration class

### ISSUE THREAD

**Issue by nsdiv** [spring-projects/spring-boot#3927]
This issue is specific to 1.3.0.* builds. We have two classes `RestClientConfig` and `RestClientSettings`. 

```java
    @Configuration
    public class RestClientConfig {

        @Autowired
        private RestClientSettings restClientSettings;

        @Bean
    	public HttpClient httpClient() {
            final PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
    		connectionManager.setDefaultMaxPerRoute(restClientSettings.getMaxConnectionsPerHost());
    		connectionManager.setMaxTotal(restClientSettings.getMaxTotalConnections());
            HttpClientBuilder httpClientBuilder = HttpClients.custom().setConnectionManager(connectionManager);
            httpClientBuilder.addInterceptorFirst(new TidHttpRequestInterceptor());
            httpClientBuilder.setRoutePlanner(new SystemDefaultRoutePlanner(ProxySelector.getDefault()));
            final CloseableHttpClient httpClient = httpClientBuilder.build();
    		return httpClient;
    	}
    }
```

```java
    @Configuration 
    @EnableConfigurationProperties
    @ConfigurationProperties(prefix="rest")
    public class RestClientSettings {
        private int connectionTimeoutMillis;
        private int readTimeoutMillis;
        private int maxConnectionsPerHost;
        private int maxTotalConnections;

        public int getConnectionTimeoutMillis() {
            return connectionTimeoutMillis;
        }

        public void setConnectionTimeoutMillis(int connectionTimeoutMillis) {
            this.connectionTimeoutMillis = connectionTimeoutMillis;
        }

        public int getReadTimeoutMillis() {
            return readTimeoutMillis;
        }

        public void setReadTimeoutMillis(int readTimeoutMillis) {
            this.readTimeoutMillis = readTimeoutMillis;
        }

        public int getMaxConnectionsPerHost() {
            return maxConnectionsPerHost;
        }

        public void setMaxConnectionsPerHost(int maxConnectionsPerHost) {
            this.maxConnectionsPerHost = maxConnectionsPerHost;
        }

        public int getMaxTotalConnections() {
            return maxTotalConnections;
        }

        public void setMaxTotalConnections(int maxTotalConnections) {
            this.maxTotalConnections = maxTotalConnections;
        }
    }
```

Using spring-boot 1.2.5, RestClientSettings was loaded correctly, with values being read properly from the application.yml file. However, after switching to 1.3.0.M5, the values are not read from the application.yml file. values for maxConnectionsPerHost, etc are coming back as 0.

Has something changed w.r.t. where the application.yml file should be located?

**Comment by wilkinsona** [comment]
No, nothing should have changed. Can you please share a complete sample that reproduces the problem?

**Comment by snicoll** [comment]
On a related note, that's not how `@ConfigurationProperties` should be used ideally. You should have one POJO with only the `@ConfigurationProperties` annotation and then in a configuration class of yours `@EnableConfigurationProperties(RestClientSettings.class)`.

**Comment by nsdiv** [comment]
@wilkinsona - Still working on getting a sample.
@snicoll - Do you mean we should only have one class that binds to external properties? We have 100s of properties that are read by 30+ classes, putting all of them in one class does not make sense.

**Comment by snicoll** [comment]
No i don't mean that. Don't put the processing of RestClientSettings on itself. Looks weird. 

It was just a remark anyway.

**Comment by nsdiv** [comment]
@wilkinsona - Here is a sample https://app.box.com/s/44x6kiph9auubyj8dr58hgv9lth04hop
Note - the pom.xml has spring-boot version set to 1.2.3.RELEASE and the test passes.
Change the spring-boot version to 1.3.0.M5 and the test fails while loading application context.

Hope this helps.

**Comment by snicoll** [comment]
If I move your configuration outside the `default` profile document then everything works fine. So I guess we broke the `default` profile loading somehow?

**Comment by snicoll** [comment]
It's a combination of two things that are mixed together.

Your test is using `@ActiveProfiles("test")`. According to the Javadoc, it is specifying the profile(s) that should run for the test and it should not merge any configuration set in configuration (see #1469). Then you are using the profile `default` that is a special name: it's the profile that should be loaded if no other profile is active. We've introduced that in the 1.3 line (#1219).

The bottom line is that your setup is incorrect, `@ActiveProfiles` should define _all_ profiles that should be active, and not expect `spring.profiles.active` to be taken into account in such case.
